[modal] Create basic ModalLauncher component

matchu opened the pull request:
In this change, we create the basic `ModalLauncher` component. We're not targeting full functionality yet; just the basics of state management, mounting, and visual overlay.

I broke these responsibilities into two components: the `ModalLauncher` itself, which handles opened state and communication with the caller; and `ModalLauncherPortal`, which handles mounting the `modal` into the DOM. Both have decent unit tests - I learned some Enzyme today, hooray!

I also installed the `flow-typed` definition for Jest, in order to be able to Flow-check our unit test files. (I also updated button's test file, and removed a `globals` comment that seems to perhaps be old and no longer needed - was that okay?)

I also started using per-component doc files, like John's doing for the grid component. This is really nice for Styleguidist, but has the downside of not being captured by our snapshot testing tool. How do y'all feel about that?

https://khanacademy.atlassian.net/browse/WB-121

Test Plan:
Unit tests pass!

Here's the full Styleguidist page:
![image](https://user-images.githubusercontent.com/59218/39326113-dcc274f2-4948-11e8-82bd-5f8222981778.png)

And here's a GIF of launching the modal:
![Styleguidist screenshot](https://phabricator-files.khanacademy.org/file/data/a35ka7ytdkiu23drzpi5/PHID-FILE-uqoifkijpibsjx6u7oby/Apr-24-2018_16-12-51.gif)

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.js:
```diff
@@ -0,0 +1,80 @@
+        // accept `React.Node` instead of `React.Element`.
+        //
+        // NOTE(mdr): Once we upgrade to React 16, we'll no longer need this
+        //     wrapper node!
```
The whole `ModalLauncherPortal` component can go away with React 16, right?

matchu replied:
Not quite, unfortunately. Built-in portals enable us to use an existing node as our render root, but don't handle creating/removing the node for you. So, we'd still have to manage the destination node ourselves.

The difference with built-in portals would be that, instead of using `ReactDOM` methods in mount/unmount, we could return a `ReactDOM.createPortal()` from `render` instead.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.test.js:
```diff
@@ -0,0 +1,57 @@
+
+        const wrapper = mount(
+            <ModalLauncherPortal>
+                <div ref={childrenRef} />
```
Interesting use of the `ref` callback.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.test.js:
```diff
@@ -0,0 +1,57 @@
+                let parent = element.parentNode;
+                while (
+                    parent.hasAttribute("data-modal-launcher-portal") ||
+                    parent.hasAttribute("data-reactroot")
```
You could use `querySelector` to find this element and then use `.contains()` to check if the root contains `element`.

matchu replied:
If I'm understanding right, I don't think that's quite the same assertion: I want to confirm that this element is mounted _directly_ into the body, modulo some expected wrapper elements. If we just asserted that the element is a descendant of `document.body`, then this test would probably pass even without portaling.

Still, there might be a clearer way to phrase this assertion, or at least leave clearer comments? I'll give it a readability pass.

matchu replied:
Oh, I misunderstood your comment! Right, I'm also not trying to assert that there _is_ a wrapper element and that this is a descendant of it; I'm trying to assert that this is mounted as near to the body as possible.

matchu replied:
I've tried adding a few more comments to clarify the intent.

kevinbarabash replied:
But if there's only one element that has `data-reactroot` and `data-modal-launcher-portal` then my approach would work.  Since that's a pretty big if, safer to go with your approach.

matchu replied:
I'm not sure that's strictly true? Like, I think the following implementation would pass the alternate version of the test:

```
render() {
    return <div data-modal-launcher-portal>{this.props.children}</div>;
}
```

If I'm understanding correctly, we'd search for the `data-modal-launcher-portal` element, check whether the children are in it, and they are so it would pass. But this component doesn't actually portal the content to the root of the page; it just renders it inline.

kevinbarabash commented on packages/wonder-blocks-modal/generated-snapshot.test.js:
```diff
@@ -58,10 +59,79 @@ describe("wonder-blocks-modal", () => {
+            />
+        );
+
+        // TODO(mdr): Use Wonder Blocks Button.
```
It grabbed your comment too, lol.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,114 @@
+        });
+
+        return (
+            <View>
```
Can we use `React.Fragment` instead.  This will ensure that the `<View>` doesn't interfere with any styling that's applied to `renderedChildren`?

matchu replied:
ooh right, will see!

matchu replied:
Looks like this won't be available until React 16; in React 15 and below, `render` must return a single element. `React.createFragment` exists to help you manage keys, but it isn't valid to return a React 15 fragment from `render`.

I'll add a comment to consider revisiting once we upgrade.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,114 @@
+        paddingLeft: 32,
+        paddingRight: 32,
+        paddingTop: 72,
+        paddingBottom: 72,
```
We should set a percentage for the width/height directly in the modals themselves.

matchu replied:
I'm still not sure about setting percentages directly in the modals, since we expect to embed them directly in layouts too, but I'll try updating this to have the correct dimensions! (Asking Tabitha some more questions in Zeplin to try to get more clarity around behavior.)

kevinbarabash replied:
The width/height can always be override by the by passing an Aphrodite style.  I think the most common use case will be using the modals with the ModalLauncher.  As long as they can be styled appropriately in the inline case it should be fine.

kevinbarabash replied:
Maybe we can use React's `context` object to know if a modal is inside a `ModalLauncherPortal` and then set things appropriately.

matchu replied:
I've changed my mind on this one. A call site might also just as easily want the "I'm previewing a modal" behavior, and want the default proportional sizing. Since I can't really predict the breadth of use cases here, I'm gonna do the simple way you suggested of inlining the rules into the modal layout component itself :) thanks for pushing on it!

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,114 @@
+ *
+ * <ModalLauncher modal={<TwoColumnModal ... />}>
+ *     {({openModal}) => <button onClick={openModal}>Learn more</button>}
+ * </ModalLauncher>
```
This is nice having an example so close to the top.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,114 @@
+        justifyContent: "center",
+
+        background: Color.offBlack64,
+    },
```
`zIndex: 1080`

matchu replied:
oooh good catch

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.test.js:
```diff
@@ -0,0 +1,60 @@
+        //     actually _mount_, which means we'd need to do a full DOM render
+        //     including `ModalLauncherPortal`, and that seems to be going
+        //     beyond the scope of this test. Really we just want to check that
+        //     this function receives a `closeModal` argument that works.
```
`enzyme` has an option to do a full render if we wanted to.  I think the test you have are sufficient though.

matchu replied:
Yeah, my concern here isn't really to do with feasibility so much as keeping the test well-scoped. It would be odd for a bug in `ModalLauncherPortal` to trigger a failure in the "modal contents can self-close" test. It might be appropriate to have an integration test that does a full render like that, but clearly labeled as an integration test?

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.test.js:
```diff
@@ -0,0 +1,60 @@
+        //     this function receives a `closeModal` argument that works.
+        const modalFn = ({closeModal}) => {
+            expect(opened).toBe(true);
+            setImmediate(closeModal);
```
Can you add a comment about why `setImmediate` is necessary here?

matchu replied:
Hmm good call. Mostly I'm trying to simulate more realistic behavior: `closeModal` really shouldn't be called immediately during this render phase, and in fact will trigger a complaint from React that you're calling `setState` during `render`. (Hmm, maybe I should actually explicitly disallow this, and write a test for it!)

matchu replied:
Eh, I guess the consequences of calling `setState` during `render` aren't so bad, and the error will announce itself if a dev makes that mistake. We shouldn't do it in our test, and I'll explain that (and maybe leave a comment), but explicitly disallowing it is probably excessive.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/index.test.js:
```diff
@@ -0,0 +1,60 @@
+        // Mount the modal launcher. This shouldn't trigger any closing yet,
+        // because we shouldn't be calling the `modal` function yet.
+        const wrapper = shallow(
+            <ModalLauncher modal={modalFn} onClose={onClose}>
```
I wonder if we should have `onClose` on the modal itself?  The reasoning, if we have it on `ModalLauncher` then users may have to keep track of which modal so that they can respond appropriately.  The only reason why I put it here is b/c the ModalLauncher also needs to call something when the user taps the backdrop to dismiss the modal.  Maybe we could just grab `this.props.modal.props.onClose` and call that function from `ModalLauncher` when someone taps the backdrop?

matchu replied:
Hmm nod. I think it makes sense for `ModalLauncher` to be the canonical source of closedness, since it controls the closed state, and I would find it weird for children to be aware of that.

It sounds like the case you're describing is using one shared modal launcher to conditionally render multiple different kinds of modals? In that case, I wonder if it's more appropriate to have multiple modal launchers instead.

I wonder if the modal's `onClose` prop in the spec should be renamed `onClickCloseButton`, to clarify that it's a UI-coupled event rather than signaling changes to the modal's actual open/closed state?

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.js:
```diff
@@ -0,0 +1,80 @@
+
+        ReactDOM.unmountComponentAtNode(destination);
+
+        if (!destination.parentNode) {
```
Is this sometimes falsy after unmounting?

kevinbarabash replied:
`flow` strikes again, got it.

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,140 @@
+import ModalLauncherPortal from "./modal-launcher-portal.js";
+
+type Props = {
+    // The modal to render.
```
The format of these comments aren't right and won't show up in styleguidist, you'll need to use `/** ... */` style comments

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,140 @@
+    //
+    // Note: Don't call `closeModal` while rendering! It should be used to
+    // respond to user intearction, like `onClick`.
+    modal: React.Node | (({closeModal: () => void}) => React.Node),
```
I'm wondering if we should ever allow devs to specify the modal directly, instead of a function. Specifying it directly is almost always an anti-pattern as it means that we're not async-loading the modal, whereas with the callback we could be. Another option that we may want to consider here is something like: `modal={() => import("...")}` for handling the most common case.

jeresig replied:
If we're expecting devs to use this component directly then we should be trying to get them to async load the modals, if at all possible. No point in having to load all of the modal-related logic if we're just waiting for the loader to activate it, right?

matchu replied:
Hmm makes sense. @kevinbarabash what do you think?

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,140 @@
+
+    _closeModal = () => {
+        this.setState({opened: false}, () => {
+            this.props.onClose && this.props.onClose();
```
I feel like I'm never going to win the fight to do:

```
if (this.props.onClose) {
    this.props.onClose();
}
```

I could've sworn we had something in our style guide but I don't see it now. Oh well!

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,140 @@
+        position: "fixed",
+        left: 0,
+        top: 0,
+        zIndex: 1080, // NOTE(mdr): Copied from Khan Academy webapp.
```
This feels like the sort of thing that we're going to want to have in a constants file of some sort - maybe insider of wonder-blocks-core.

kevinbarabash replied:
Eventually, I'd like to have a way to inject constants so that we don't have to maintain constants in two different codebases.  It also supports re-use of these components by other people outside the org.

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.js:
```diff
@@ -0,0 +1,80 @@
+        return root;
+    }
+
+    render() {
```
Curious: can you just remove the render method? Is React ok with that?

jeresig commented on packages/wonder-blocks-modal/docs.md:
````diff
@@ -54,3 +54,67 @@ const styles = StyleSheet.create({
     />
 </View>;
 ```
+
````
Hmm, rather than adding all of these to a single docs.md file it'd be better to add this to a `modal-launcher.md` file (that way it's shown with the modal-launcher docs). Same goes for the two-column modal.

matchu replied:
Hmm nod, the reason I was avoiding this was because snapshot tests didn't pick them up, but maybe that's a quick fix we should just do.

kevinbarabash replied:
We need to update `gen-snapshot-tests.js` to handle that case.

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/modal-launcher-portal.js:
```diff
@@ -0,0 +1,80 @@
+// @flow
+/**
```
Are we expecting users to ever use this component? It seems like we should make that explicit here in the comments.

matchu replied:
Hmm nod, I wonder if maybe it should even be hidden from Styleguidist somehow? I don't think I even want to mention this component to consumers. Could `components/private/modal-launcher-portal.js` make sense?

jeresig replied:
You can explicitly hide it from styleguidist! I think this is the right way: https://react-styleguidist.js.org/docs/cookbook.html#how-to-exclude-some-components-from-style-guide

matchu replied:
Hmm nod, I wonder if maybe it should even be hidden from Styleguidist somehow? I don't think I even want to mention this component to consumers. Could `components/private/modal-launcher-portal.js` make sense?

matchu replied:
Hm okay! It feels weird to do that at a global level, but I suppose that's not so bad since this will probably be uncommon.

jeresig replied:
Oh - you don't have to do it on the global level, you can include the ignore option at the section level, too: https://react-styleguidist.js.org/docs/components.html#sections

matchu replied:
Oh neat, thanks!

kevinbarabash replied:
Nice!

jeresig commented on packages/wonder-blocks-modal/components/modal-launcher/index.js:
```diff
@@ -0,0 +1,140 @@
+// @flow
```
hmm... I'm not a fan of the file naming of this component I think I'd prefer to just have all the js files be inside of the components directory. And if you really want this to be in a sub-directory then rename this file to be `modal-launcher.js` and just make the `index.js` file: `export * from "modal-launcher.js";`, like we've done elsewhere.

kevinbarabash replied:
Same.  I find it hard to differentiate files when there are multiple `index.js`s in a codebase.

matchu commented:
Lotsa changes in the works! Might need a re-review from y'all in a bit; still moving some pieces around.

matchu commented:
Okay, I think this version captures y'all's main concerns! Could y'all take another look?

Major changes include:
* The `modal-launcher` subdirectory is gone; its contents have been flattened into `components`.
* I refactored the components' inline doc comments to be correctly captured by Styleguidist.
* I split `docs.md` into `modal-launcher.md` and `two-column-modal.md`, which read much nicer in Styleguidist, but aren't captured by snapshot testing under our current tooling. Is that okay?

Thanks!

matchu commented:
👍🏻 I've logged the async question as a comment in the spec.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher.js:
```diff
@@ -0,0 +1,141 @@
+        });
+
+        return (
+            // TODO(mdr): This can be a fragment once we're on React 16.
```
I keep forgetting that that's a React 16 thing.  All the more reason to upgrade sooner rather than later.

kevinbarabash commented on packages/wonder-blocks-modal/components/modal-launcher.js:
```diff
@@ -0,0 +1,141 @@
+        // TODO(mdr): Should we keep this in a constants file somewhere? Or
+        //     not hardcode it at all, and provide it to Wonder Blocks via
+        //     configuration?
+        zIndex: 1080,
```
Configuration, but that can happen in a future PR.

kevinbarabash commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -33,6 +36,13 @@ const styles = StyleSheet.create({
         alignItems: "stretch",
+
+        width: "86.72%",
+        height: "60.42%",
```
thanks
